### DISCUSSION TITLE
Wrong rendering of island/areas near water - Geometry issue of water data?

### DISCUSSION THREAD

**Discussion by christian2701** [maplibre/maplibre-native#1836] [Q&A]
Hello everyone, we are currently working on a conversion of our stack to Maplibre and would also like to provide the raster tiles via Tileserver/Maplibre Native.

Unfortunately, there is sometimes a server crash/wrong rendering on islands/areas near water at certain zoom levels.

Here is an example in the Venice area for 15/17506/11729.webp. For test purposes I have coloured the corresponding water area yellow:

![11729](https://github.com/maplibre/maplibre-native/assets/48522137/f0613c12-dad7-4572-825d-e3ba53493c05)

At "raster" zoom level 14 the data is drawn in correctly:

![correct](https://github.com/maplibre/maplibre-native/assets/48522137/b722afc4-4ec2-4f68-a686-1ba7773ec730)

I also have attached the corresponding mbiltes (please remove .txt at the end):

[it-nord-est-latest-wateronly-lowres.mbtiles.txt](https://github.com/maplibre/maplibre-native/files/13251477/it-nord-est-latest-wateronly-lowres.mbtiles.txt)

And the data in geojson format (please remove .txt at the end):
[5864.geojson.txt](https://github.com/maplibre/maplibre-native/files/13251589/5864.geojson.txt)

The problem are wrong generated water geometries. These are generated as follows (see also shortbread on which the tiles are based https://github.com/shortbread-tiles/shortbread-tilemaker/blob/main/process.lua#L378-419):

```lua
function process_water_polygons(way, way_area)
	local waterway = way:Find("waterway")
	local natural = way:Find("natural")
	local water = way:Find("water")
	local landuse = way:Find("landuse")
	local mz = inf_zoom
	local kind = ""
	local is_river = (natural == "water" and water == "river") or waterway == "riverbank"
	if landuse == "reservoir" or landuse == "basin" or (natural == "water" and not is_river) or natural == "glacier" then
		mz = math.max(4, zmin_for_area(way, 0.01, way_area))
		if mz >= 10 then
			mz = math.max(10, zmin_for_area(way, 0.1, way_area))
		end
		if landuse == "reservoir" or landuse == "basin" then
			kind = landuse
		elseif natural == "water" or natural == "glacier" then
			kind = natural
		end
	elseif is_river or waterway == "dock" or waterway == "canal" then
		mz = math.max(4, zmin_for_area(way, 0.1, way_area))
		kind = waterway
		if kind == "" then
			kind = water
		end
	end
	if mz < inf_zoom then
		local way_area = way_area
		way:Layer("water_polygons", true)
		way:MinZoom(mz)
		way:Attribute("kind", kind)
		way:AttributeNumeric("way_area", way_area)
		way:ZOrder(way_area)
		if way:Holds("name") then
			way:LayerAsCentroid("water_polygons_labels")
			-- way:MinZoom(13)
			way:MinZoom(10)
			way:Attribute("kind", kind)
			way:AttributeNumeric("way_area", way_area)
			way:ZOrder(way_area)
			setNameAttributes(way)
		end
	end
end
```

The corresponding style for this layer:

```json
{
  "id": "water-area",
  "type": "fill",
  "source": "smartmaps",
  "source-layer": "water_polygons",
  "minzoom": 12,
  "filter": [
    "all",
    ["==", "kind", "water"]
  ],
  "layout": {"visibility": "visible"},
  "paint": {
    "fill-color": "#a2c1f7",
    "fill-opacity": 1
  }
}
```

I have noticed that the water geometries created here sometimes look strange. Besides the actual geometry, a "border" in the upper right corner is created to close the geometry:

![wrong_geometry](https://github.com/maplibre/maplibre-native/assets/48522137/4cf5f62f-42da-4239-8ea2-cccd46d0a9f9)

In comparison the OpenMapTiles geometry looks like this and is rendered correct:
![omt_correct_geometry](https://github.com/maplibre/maplibre-native/assets/48522137/bb1c4e29-38fb-41c6-aebf-d4862c6cb52c)

In "vector mode" there isn't any issue, the problem only occurs when using the tile mode / raster images.

Do you have any ideas or suggestions for analysing the problem in more detail? Thank you very much!

Note: I already have mentioned the issue in the tileserver-repo, but I think it could be a maplibre-native issue.

**Comment by acalcutt** [comment]
I was wondering if you have tried to replicate at all with the core [mbgl-render](https://github.com/maplibre/maplibre-native/tree/main/platform/linux#mbgl-render) utility . Sometimes that is a good alternative to tileserver-gl to see if this is a maplibre-native core issue.

For example, in a previous PR i did something like
`./build/bin/mbgl-render --style style.json --output out.png --zoom 6 --lat 39.470 --lon -103`

**Reply by acalcutt** [reply]
I think I was able to confirm your issue shows in maplibre-native core with mbgl-render

I took the example mbtiles files you gave and combined it with your style layer to show water only and made this test style
[test.zip](https://github.com/maplibre/maplibre-native/files/13257518/test.zip)

then with that I ran 
`xvfb-run -a /opt/test7/mbntest/build/bin/mbgl-render --style=style.json --lat=45.446518 --lon=12.330183 --zoom=16 --width=4000 --height=4000 --output 4000.png`

and ended up with this image
![4000](https://github.com/maplibre/maplibre-native/assets/3792408/d7a3304c-be0e-403c-9851-bb895e2818b8)

You can see in that there is a tile in there that is all water, like i observed in https://github.com/maptiler/tileserver-gl/issues/1041#issuecomment-1793488591 with tileserver.

**Reply by acalcutt** [reply]
but like you said, with maplibre-js and inspect this area looks fine
![image](https://github.com/maplibre/maplibre-native/assets/3792408/feb28325-f8ae-4fe7-b881-1beb14fa1cb5)

I also looked at your pbf in qgis and it looked ok to me

**Reply by christian2701** [reply]
@acalcutt Thank you Andrew for further analysing. We are still unsure where exactly the problem lies, we get the same error in Key Largo or in Copenhagen (vector is fine everywhere). We will continue to try to solve the problem and are very grateful for any further help in the meantime.

**Comment by geoneutrino** [comment]
Looks to me like a problem in maplibre-native processing. Changing fill-opacity from 1 to e.g. 0.4 shows all geometries are there and rendered but "something" is painted over the others
Hm, maybe as maplibre-js as well as Qgis work correctly some edge case in geometry processing or such ?

![4000-2](https://github.com/maplibre/maplibre-native/assets/48556918/f851aa15-3d80-4c9c-956b-26efee176f80)
